dcp-0006/dcp-0006.mediawiki

```diff
@@ -0,0 +1,395 @@
+# If the TSpend transaction does not make it into a block on the final TVI in the window the Expiry will force it from the mempool on the next block.
+
+A TSpend transaction voting window and validity is completely calculated from
+the Expiry field. The Expiry shall always be any TVI+2. This odd number is
+emergent from the handling of Expiry in the mempool. A transaction is expired
+from the Mempool one block prior to the actual Expiry. Thus in order for the
+Tspend transaction to not get evicted from the mempool on the last possible TVI
+the Expiry must be set to +2.
+
+The start of a TSpend transaction voting window is calculated as follows: Expiry -
+(TVI*TVIM) - 2.
+The end of a TSpend transaction voting window is calculated as follows: Expiry
+- 2.
+A TSpend transaction is considered "inside the window" when the block height >=
+start and block height <= end. The end parameter is inclusive for allowing a
+TSpend in a block however it is exclusive for allowing votes on the TSpend that
+may end up being allowed on the block.
+
+A treasury spend transaction is identified by the following characteristics:
+
+* The wire message transaction version shall be 3.
+* There shall be one input script.
+* There shall be at least two output scripts.
+* All output scripts shall be version 0.
+* TxIn[0] signature script shall be exactly 100 bytes in length. The script format is <code>OP_DATA_64</code> (signature), <code>OP_DATA_33</code> (Pi public key) followed by an <code>OP_TSPEND</code>.
+** The Pi public key that is included in TxIn[0] shall be of the strict compressed public key encoding type.
+* TxOut[0] script shall be an <code>OP_RETURN</code> <code>OP_DATA_32</code>.
+* TxOut[1:N] scripts shall be an <code>OP_TGEN</code> followed by a P2PKH or a P2SH script.
+
+{|
+!Transaction!!Script!!Description
+|-
+|TxIn[0]||[0]OP_DATA_64 signature<br>[65]OP_DATA_33 Pi public key<br>[99]OP_TSPEND||The value shall be sum(expenditures) + mining fee.
+|-
+|TxOut[0]||[0]OP_RETURN<br>[1]OP_DATA_32||Bytes that make the entire transaction unique. First 8 bytes are the little endian encoded value from TxIn[0]. The following 24 bytes are random.
+|-
+|TxOut[1:N]||[0]OP_TGEN<br>[1:M]P2PKH or P2SH script||These are the individual payouts from the treasury account.
+|}
+
+Additional consensus enforced characteristics:
+* The previous out point for TxIn[0] shall be null.
+* The TSpend transaction fraud proof shall be null.
+* The length of the signature script in TxIn[0] shall be 100 bytes ([OP_DATA_64] [signature] [OP_DATA_33] [PiKey] [OP_TSPEND] = 1 + 64 + 1 + 33 + 1 = 100)
+* A TSpend transaction shall not be allowed in a block that is not a TVI.
+* A TSpend transaction shall not be allowed in a block if the start of the voting window is prior to stake validation height.
+* A TSpend transaction shall not be allowed in a block if the Expiry is outside of the TSpend validity window.
```

Comment:
"voting window and validity is completely calculated from the Expiry field" earlier suggests validity is dictated by Expiry, but this reads like the other way round. Which one is true?

Reply:
This is exactly true. I don't know how to make that more obvious than how it is written tbh.

Reply:
Well if "TSpend validity window" is a function of Expiry, not allowing Expiry outside of same window is a circular reference.

Perhaps it refers to "not allowed ... if the Expiry is farther than 2*TVI*TVIM block height away" in the updated version?